Delegate method for errors during Bonjour registration

IPv6 opened the pull request:
We are using your server in bonjour-centered app and this delegate method is needed to know about possible errors during server bootstrap

lucaji commented:
for sake of correctness it is correct "didFailBonjourRegistration" and not "didFailed".

swisspol commented on GCDWebServer/Core/GCDWebServer.h:
```diff
@@ -217,6 +217,10 @@ extern NSString* const GCDWebServerAuthenticationMethod_DigestAccess;
 - (void)webServerDidCompleteBonjourRegistration:(GCDWebServer*)server;
 
 /**
+ *  This method is called if Bonjour registration were failed for any reason */
```
Should be "This method is called if the Bonjour registration fails for any reason." and the `*/` should be on the next line like the other documentation entries.

swisspol commented on GCDWebServer/Core/GCDWebServer.h:
```diff
@@ -217,6 +217,10 @@ extern NSString* const GCDWebServerAuthenticationMethod_DigestAccess;
 
 /**
+ *  This method is called if Bonjour registration were failed for any reason */
+- (void)webServerDidFailedBonjourRegistration:(GCDWebServer*)server withError:(NSError*)error;
```
`-webServerDidFailBonjourRegistration:withError:`

swisspol commented on GCDWebServer/Core/GCDWebServer.m:
```diff
@@ -564,14 +570,19 @@ - (BOOL)_start:(NSError**)error {
       CFNetServiceSetClient(_registrationService, _NetServiceRegisterCallBack, &context);
       CFNetServiceScheduleWithRunLoop(_registrationService, CFRunLoopGetMain(), kCFRunLoopCommonModes);
-      CFStreamError streamError = {0};
-      CFNetServiceRegisterWithOptions(_registrationService, 0, &streamError);
```
Don't change the order of the calls.

IPv6 commented:
i fixed names/comments mentioned, sorry for mistakes. i also added delegate method for tracking exact moment of request handling finish with error code. may be it`s not ok to use http codes, but we have to see when writeBody fails (sending big files frequently stops with socket error "Broken pipe" on low-bandwidth netwroks, we have to notify user about network conditions in this case)

i undestand that may be all this mini-changes not so useful (but we need them in our app), please tell so i will not bother you with such things. i just decided to participate here a bit, GCDWebServer is good and easy-to-use project )

swisspol commented:
Thanks for the update, but it's not a pull request I'm going to be able to accept: it just has too many changes.

You don't need to add a delegate method for HTTP errors (which won't be called on the main thread in your current implementation by the way), instead I'd recommend using a custom subclass of `GCDWebServerConnection` with the `GCDWebServerOption_ConnectionClass` option. Then override the methods you want.

IPv6 commented:
thank you, got it, you are right.

BTW, about calling delegate from main thread, there is one caveat i already stumbled with GCDWebServer:
- webServerDidStop called in main thread
- server stops each time user closes app
- when app is going into background delegate call block is scheduled, but NOT called, due transition to background. and already scheduled call is called when app is going back to foreground (and server starts again)

its confusing / i understand that this is already established call convention, but webServerDidStop became unreliable this way

swisspol commented:
Thanks for reporting this issue, I created a separate one #144.
